File: backend/core/payout_schedule.py
"""
Payout schedule helpers for TradeSafe release windows.
"""

from __future__ import annotations

from datetime import datetime, timedelta, timezone
from typing import Iterable, List


DEFAULT_TIMEZONE_OFFSET_HOURS = 2


def _sa_now(now: datetime | None = None) -> datetime:
    current = now or datetime.now(timezone.utc)
    if current.tzinfo is None:
        current = current.replace(tzinfo=timezone.utc)
    return current.astimezone(timezone(timedelta(hours=DEFAULT_TIMEZONE_OFFSET_HOURS)))


def _parse_hhmm(value: str) -> tuple[int, int]:
    hours, minutes = value.split(":", 1)
    return int(hours), int(minutes)


def _easter_sunday(year: int) -> datetime:
    a = year % 19
    b = year // 100
    c = year % 100
    d = b // 4
    e = b % 4
    f = (b + 8) // 25
    g = (b - f + 1) // 3
    h = (19 * a + b - d - g + 15) % 30
    i = c // 4
    k = c % 4
    l = (32 + 2 * e + 2 * i - h - k) % 7
    m = (a + 11 * h + 22 * l) // 451
    month = (h + l - 7 * m + 114) // 31
    day = ((h + l - 7 * m + 114) % 31) + 1
    return datetime(year, month, day, tzinfo=timezone(timedelta(hours=DEFAULT_TIMEZONE_OFFSET_HOURS)))


def get_south_african_public_holidays(year: int) -> set[str]:
    easter = _easter_sunday(year)
    fixed_dates = [
        (1, 1),   # New Year's Day
        (3, 21),  # Human Rights Day
        (4, 27),  # Freedom Day
        (5, 1),   # Workers' Day
        (6, 16),  # Youth Day
        (8, 9),   # National Women's Day
        (9, 24),  # Heritage Day
        (12, 16), # Day of Reconciliation
        (12, 25), # Christmas Day
        (12, 26), # Day of Goodwill
    ]

    holiday_dates = {f"{year}-{month:02d}-{day:02d}" for month, day in fixed_dates}
    holiday_dates.add((easter - timedelta(days=2)).date().isoformat())
    holiday_dates.add((easter + timedelta(days=1)).date().isoformat())

    observed = set()
    for item in list(holiday_dates):
        holiday_date = datetime.fromisoformat(item).replace(tzinfo=timezone(timedelta(hours=DEFAULT_TIMEZONE_OFFSET_HOURS)))
        if holiday_date.weekday() == 6:  # Sunday
            observed.add((holiday_date + timedelta(days=1)).date().isoformat())

    return holiday_dates | observed


def is_business_day(date: datetime) -> bool:
    local = _sa_now(date)
    if local.weekday() >= 5:
        return False
    holidays = get_south_african_public_holidays(local.year)
    return local.date().isoformat() not in holidays


def next_business_day(date: datetime) -> datetime:
    local = _sa_now(date)
    next_day = local
    while True:
        next_day = next_day + timedelta(days=1)
        if is_business_day(next_day):
            return next_day


def _normalize_times(times: Iterable[str] | None, fallback: List[str]) -> List[str]:
    values = list(times or fallback)
    return [time for time in values if time]


def get_next_payout_release(
    now: datetime | None = None,
    release_times: Iterable[str] | None = None,
    cutoff_times: Iterable[str] | None = None,
) -> dict:
    """
    Calculate the next payout release window using South African business-day rules.
    """
    local_now = _sa_now(now)
    releases = _normalize_times(release_times, ["10:00", "15:00"])
    cutoffs = _normalize_times(cutoff_times, ["09:00", "14:00"])

    release_10 = _parse_hhmm(releases[0])
    release_15 = _parse_hhmm(releases[1] if len(releases) > 1 else releases[0])
    cutoff_10 = _parse_hhmm(cutoffs[0])
    cutoff_15 = _parse_hhmm(cutoffs[1] if len(cutoffs) > 1 else cutoffs[0])

    current_minutes = local_now.hour * 60 + local_now.minute
    cutoff_10_minutes = cutoff_10[0] * 60 + cutoff_10[1]
    cutoff_15_minutes = cutoff_15[0] * 60 + cutoff_15[1]

    if not is_business_day(local_now):
        target_date = next_business_day(local_now)
        target = target_date.replace(hour=release_10[0], minute=release_10[1], second=0, microsecond=0)
    elif current_minutes < cutoff_10_minutes:
        target = local_now.replace(hour=release_10[0], minute=release_10[1], second=0, microsecond=0)
    elif current_minutes < cutoff_15_minutes:
        target = local_now.replace(hour=release_15[0], minute=release_15[1], second=0, microsecond=0)
    else:
        target_date = next_business_day(local_now)
        target = target_date.replace(hour=release_10[0], minute=release_10[1], second=0, microsecond=0)

    if target <= local_now:
        target = target + timedelta(days=1)
        while not is_business_day(target):
            target = target + timedelta(days=1)
        target = target.replace(hour=release_10[0], minute=release_10[1], second=0, microsecond=0)

    return {
        "release_at": target.isoformat(),
        "label": format_release_label(target, local_now),
        "business_day": is_business_day(local_now),
        "timezone": "Africa/Johannesburg",
        "release_times": releases,
        "cutoff_times": cutoffs,
    }


def format_release_label(target: datetime, now: datetime | None = None) -> str:
    local_now = _sa_now(now)
    local_target = _sa_now(target)
    days_ahead = (local_target.date() - local_now.date()).days

    if days_ahead <= 0:
        prefix = "Today"
    elif days_ahead == 1:
        prefix = "Tomorrow"
    elif days_ahead < 7:
        prefix = local_target.strftime("%A")
    else:
        prefix = local_target.strftime("%d %b")

    return f"{prefix} {local_target.strftime('%H:%M')}"


def build_payout_schedule_summary(
    now: datetime | None = None,
    *,
    release_times: Iterable[str] | None = None,
    cutoff_times: Iterable[str] | None = None,
    clearing_disclaimer: str = "Bank clearing may take up to 2 business days depending on payment runs, weekends, and bank processing.",
) -> dict:
    next_release = get_next_payout_release(now, release_times=release_times, cutoff_times=cutoff_times)
    if not next_release["business_day"]:
        short_label = "Weekend payout: next business day"
    else:
        short_label = f"Next payout release: {next_release['label']}"

    return {
        "next_release": next_release,
        "short_label": short_label,
        "copy": f"{short_label}. {clearing_disclaimer}",
        "clearing_disclaimer": clearing_disclaimer,
    }

File: frontend/src/utils/payoutSchedule.js
const TIME_ZONE = 'Africa/Johannesburg';
const DEFAULT_RELEASE_TIMES = ['10:00', '15:00'];
const DEFAULT_CUTOFF_TIMES = ['09:00', '14:00'];
const DEFAULT_CLEARING_DISCLAIMER = 'Bank clearing may take up to 2 business days depending on payment runs, weekends, and bank processing.';

function formatDateParts(date) {
  const parts = new Intl.DateTimeFormat('en-ZA', {
    timeZone: TIME_ZONE,
    year: 'numeric',
    month: '2-digit',
    day: '2-digit',
    hour: '2-digit',
    minute: '2-digit',
    second: '2-digit',
    hour12: false,
  }).formatToParts(date);

  const lookup = Object.fromEntries(parts.map((part) => [part.type, part.value]));
  return {
    year: Number(lookup.year),
    month: Number(lookup.month),
    day: Number(lookup.day),
    hour: Number(lookup.hour),
    minute: Number(lookup.minute),
    second: Number(lookup.second),
  };
}

function buildSaDate({ year, month, day, hour = 0, minute = 0, second = 0 }) {
  return new Date(Date.UTC(year, month - 1, day, hour - 2, minute, second));
}

function startOfSaDay(date) {
  const parts = formatDateParts(date);
  return buildSaDate({ ...parts, hour: 0, minute: 0, second: 0 });
}

function parseHm(value) {
  const [hour, minute] = String(value).split(':').map(Number);
  return { hour, minute };
}

function easterSunday(year) {
  const a = year % 19;
  const b = Math.floor(year / 100);
  const c = year % 100;
  const d = Math.floor(b / 4);
  const e = b % 4;
  const f = Math.floor((b + 8) / 25);
  const g = Math.floor((b - f + 1) / 3);
  const h = (19 * a + b - d - g + 15) % 30;
  const i = Math.floor(c / 4);
  const k = c % 4;
  const l = (32 + 2 * e + 2 * i - h - k) % 7;
  const m = Math.floor((a + 11 * h + 22 * l) / 451);
  const month = Math.floor((h + l - 7 * m + 114) / 31);
  const day = ((h + l - 7 * m + 114) % 31) + 1;
  return { year, month, day };
}

function getHolidaySet(year) {
  const toKey = (date) => {
    const parts = formatDateParts(date);
    return `${parts.year}-${String(parts.month).padStart(2, '0')}-${String(parts.day).padStart(2, '0')}`;
  };
  const fixedDates = [
    [1, 1],
    [3, 21],
    [4, 27],
    [5, 1],
    [6, 16],
    [8, 9],
    [9, 24],
    [12, 16],
    [12, 25],
    [12, 26],
  ];

  const holidays = new Set(fixedDates.map(([month, day]) => `${year}-${String(month).padStart(2, '0')}-${String(day).padStart(2, '0')}`));
  const easter = easterSunday(year);
  const easterDate = new Date(Date.UTC(easter.year, easter.month - 1, easter.day, 12));
  holidays.add(toKey(new Date(easterDate.getTime() - 2 * 24 * 60 * 60 * 1000)));
  holidays.add(toKey(new Date(easterDate.getTime() + 1 * 24 * 60 * 60 * 1000)));

  Array.from(holidays).forEach((value) => {
    const date = new Date(`${value}T12:00:00Z`);
    if (date.getUTCDay() === 0) {
      holidays.add(toKey(new Date(date.getTime() + 24 * 60 * 60 * 1000)));
    }
  });

  return holidays;
}

function isBusinessDay(date) {
  const parts = formatDateParts(date);
  const dayKey = `${parts.year}-${String(parts.month).padStart(2, '0')}-${String(parts.day).padStart(2, '0')}`;
  const utcDay = new Date(Date.UTC(parts.year, parts.month - 1, parts.day, 12)).getUTCDay();
  if (utcDay === 0 || utcDay === 6) return false;
  return !getHolidaySet(parts.year).has(dayKey);
}

function nextBusinessDay(date) {
  let current = startOfSaDay(date);
  while (true) {
    current = new Date(current.getTime() + 24 * 60 * 60 * 1000);
    if (isBusinessDay(current)) return current;
  }
}

function localReleaseLabel(target, now) {
  const targetParts = formatDateParts(target);
  const nowParts = formatDateParts(now);
  const targetKey = `${targetParts.year}-${String(targetParts.month).padStart(2, '0')}-${String(targetParts.day).padStart(2, '0')}`;
  const nowKey = `${nowParts.year}-${String(nowParts.month).padStart(2, '0')}-${String(nowParts.day).padStart(2, '0')}`;
  let prefix = 'Today';
  if (targetKey !== nowKey) {
    const dayDiff = Math.round((startOfSaDay(target).getTime() - startOfSaDay(now).getTime()) / (24 * 60 * 60 * 1000));
    if (dayDiff === 1) prefix = 'Tomorrow';
    else prefix = new Intl.DateTimeFormat('en-ZA', { timeZone: TIME_ZONE, weekday: 'long' }).format(target);
  }
  return `${prefix} ${String(targetParts.hour).padStart(2, '0')}:${String(targetParts.minute).padStart(2, '0')}`;
}

function normalizeList(values, fallback) {
  const list = Array.isArray(values) && values.length ? values : fallback;
  return list.filter(Boolean);
}

export function getNextPayoutRelease(now = new Date(), config = {}) {
  const releaseTimes = normalizeList(config.payout_release_times, DEFAULT_RELEASE_TIMES);
  const cutoffTimes = normalizeList(config.payout_cutoff_times, DEFAULT_CUTOFF_TIMES);
  const localNow = now;
  const parts = formatDateParts(localNow);
  const minutes = parts.hour * 60 + parts.minute;
  const release10 = parseHm(releaseTimes[0]);
  const release15 = parseHm(releaseTimes[1] || releaseTimes[0]);
  const cutoff10 = parseHm(cutoffTimes[0]);
  const cutoff15 = parseHm(cutoffTimes[1] || cutoffTimes[0]);
  const cutoff10Minutes = cutoff10.hour * 60 + cutoff10.minute;
  const cutoff15Minutes = cutoff15.hour * 60 + cutoff15.minute;

  let target;
  if (!isBusinessDay(localNow)) {
    const nextDay = nextBusinessDay(localNow);
    target = buildSaDate({ ...formatDateParts(nextDay), hour: release10.hour, minute: release10.minute, second: 0 });
  } else if (minutes < cutoff10Minutes) {
    target = buildSaDate({ ...parts, hour: release10.hour, minute: release10.minute, second: 0 });
  } else if (minutes < cutoff15Minutes) {
    target = buildSaDate({ ...parts, hour: release15.hour, minute: release15.minute, second: 0 });
  } else {
    const nextDay = nextBusinessDay(localNow);
    target = buildSaDate({ ...formatDateParts(nextDay), hour: release10.hour, minute: release10.minute, second: 0 });
  }

  const isWeekendOrHoliday = !isBusinessDay(localNow);
  const releaseLabel = localReleaseLabel(target, localNow);
  const headline = isWeekendOrHoliday ? `Weekend payout: next business day` : `Next payout release: ${releaseLabel}`;

  return {
    releaseAt: target.toISOString(),
    releaseLabel,
    headline,
    isBusinessDay: !isWeekendOrHoliday,
    releaseTimes,
    cutoffTimes,
  };
}

export function getPayoutScheduleMessage(now = new Date(), config = {}) {
  const summary = getNextPayoutRelease(now, config);
  const disclaimer = config.payout_clearing_disclaimer || DEFAULT_CLEARING_DISCLAIMER;
  return {
    ...summary,
    disclaimer,
    copy: `${summary.headline}. ${disclaimer}`,
    shortCopy: summary.headline,
  };
}

export function getDefaultMinimumTransactionAmount(config = {}) {
  const value = Number(config.minimum_transaction);
  return Number.isFinite(value) && value > 0 ? value : 500;
}

export function getPayoutClearingDisclaimer(config = {}) {
  return config.payout_clearing_disclaimer || DEFAULT_CLEARING_DISCLAIMER;
}

export function calculatePayoutSchedule(releasedAt, config = {}) {
  const releasedDate = new Date(releasedAt);
  const releaseParts = formatDateParts(releasedDate);
  const releaseMinutes = releaseParts.hour * 60 + releaseParts.minute;

  const releaseTimes = normalizeList(config.payout_release_times, DEFAULT_RELEASE_TIMES);
  const release10 = parseHm(releaseTimes[0]);
  const release15 = parseHm(releaseTimes[1] || releaseTimes[0]);
  const threshold10 = release10.hour * 60 + release10.minute;
  const threshold15 = release15.hour * 60 + release15.minute;

  let payoutRunAt;

  if (!isBusinessDay(releasedDate) || releaseMinutes >= threshold15) {
    const nextBiz = nextBusinessDay(releasedDate);
    const nextNextBiz = nextBusinessDay(nextBiz);
    payoutRunAt = buildSaDate({ ...formatDateParts(nextNextBiz), hour: release10.hour, minute: release10.minute, second: 0 });
  } else if (releaseMinutes >= threshold10) {
    const nextBiz = nextBusinessDay(releasedDate);
    payoutRunAt = buildSaDate({ ...formatDateParts(nextBiz), hour: release15.hour, minute: release15.minute, second: 0 });
  } else {
    const nextBiz = nextBusinessDay(releasedDate);
    payoutRunAt = buildSaDate({ ...formatDateParts(nextBiz), hour: release10.hour, minute: release10.minute, second: 0 });
  }

  const pp = formatDateParts(payoutRunAt);
  const bankRunLabel = `${String(pp.hour).padStart(2, '0')}:${String(pp.minute).padStart(2, '0')}`;

  return { payoutRunAt, bankRunLabel };
}
